pkg/util/errors.go

```diff
@@ -66,3 +119,88 @@ func IsConnCanceled(err error) bool {
 
 	return false
 }
+
+// CodedErrorSampler keeps a sample of entries giving priority to entries which implement CodedError and have a code >= 500
+// errors which do not implement CodedError are treated the same as code == 500
+type CodedErrorSampler struct {
+	keepMax int
+	codes   []int
+	errors  []error
+}
+
+func NewCodedErrorSampler(keepMax int) *CodedErrorSampler {
+	return &CodedErrorSampler{
+		keepMax: keepMax,
+		codes:   []int{},
```

Comment:
Any particular reason to keep codes an errors in separate slices?  And not just a slice of coded errors?